### PR TITLE
File Transfer Client in Flutter

### PR DESCRIPTION
**Pull request by ApolloZhu**
## Note for Reviewer

`boardless_outside/lib/main.dart` is the main file handling the UI, the rest are just skeleton files to get it running on each specific platform.

## Changelog

- Setup flutter project for file transfer client
- Added ability to upload any files, including text files
- Copy button for plain text
- Update GitHub Actions to automatically build and deploy the web version

## Demo

- https://uwrealitylab.github.io/xrcapstone22wi-team8/outside

## Testing

Tested on Web, Android, iOS, and macOS

## Known Issues

See https://github.com/UWRealityLab/xrcapstone22wi-team8/issues/10

## File Management Architecture

In Firebase Storage (for saving files):

```
root
  - room name (e.g. test)
    - Filename (e.g. hello.png)
```

In Firebase Firestore (the database):

```
root
  - room name (e.g. test)
    # For files
    - random id
      - name: String (to display in the list, default to same as Filename)
      - ref: Filename (e.g. hello.png)
    # For plain text
    - random id
      - name: String or null (to display in the list, default to the selected file's name, okay to be null)
      - content: String (e.g. "Hello World")
      # there's no ref because the file is never uploaded
```

### Security

All files are readable and writable by everyone since we don't have private rooms, only public lobbies.

## Why Flutter, not C# and Unity?

- Flutter is cross-platform, just like C# and Unity. It can run on mobile, desktop, and web!
- Both has good support for Firebase, so we can still access files from the main Boardless Unity project
- Unity's UI Toolkit is incomplete, and doesn't have a file picker!!! On the other hand, Flutter does
   - I learned Flutter in a fraction of the time it takes to figure out how to implement a file picker in Unity

### CODE REVIEW

File: boardless_outside/lib/firebase_options.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      case TargetPlatform.android:
+        return android;
+      case TargetPlatform.iOS:
+        return ios;
+      case TargetPlatform.macOS:
+        return macos;
+    }
+
+    throw UnsupportedError(
+      'DefaultFirebaseOptions are not supported for this platform.',
```
Comment (ApolloZhu) [inline boardless_outside/lib/firebase_options.dart:33]: Firebase doesn't seem to support other platforms, should I delete files for building on Windows and Linux?

Comment (ApolloZhu) [reply boardless_outside/lib/firebase_options.dart:33]: Folders removed

**Review by KH-Kyle — APPROVED** [review]
Most files are configurations. No script needs to be checked for styling. Should the demo page be blank?

**Comment by ApolloZhu** [comment]
> Should the demo page be blank?

@KH-Kyle the page does take a long time to load, but shouldn't be blank after about 5 to 10 seconds (an unfortunate thing with Flutter). Could you please double-check to confirm?

**Comment by KH-Kyle** [comment]
> > Should the demo page be blank?
> 
> @KH-Kyle the page does take a long time to load, but shouldn't be blank after about 5 to 10 seconds (an unfortunate thing with Flutter). Could you please double-check to confirm?

It showed up! I used Safari at first which took really long but it's 5-10 sec for Chrome.

**Comment by ApolloZhu** [comment]
Sounds good, merging

**Pull request merged**